packages/p2p/source/downloader/message-downloader.ts

```diff
@@ -130,16 +130,72 @@ export class MessageDownloader implements Contracts.P2P.Downloader {
 		try {
 			const result = await this.communicator.getMessages(job.peer);
 
-			for (const prevoteBuffer of result.prevotes) {
-				// TODO: handle response
-				const prevote = await this.factory.makePrevoteFromBytes(prevoteBuffer);
-				await this.prevoteProcessor.process(prevote, false);
+			const prevotes: Map<number, Contracts.Crypto.IPrevote> = new Map();
```

Comment:
Right now precommits / prevotes are deserialized first (loop 1) and then processed (loop 2).

So in total we have 4 loops.

We could combine the deserialize loop and the process loop so we just have 2 loops? 

An additional benefit would be that we fail early in case the height/round mismatches without having deserialized all messages beforehand which then are discarded anyway.